### ISSUE TITLE
Avoid reparsing on solution open

### ISSUE THREAD

**Issue by panopticoncentral** [dotnet/project-system#6379]
From @jasonmalinowski 

This is following up to the idea we had to have the project system pass in an approximation of the options that impact parsing once we’ve completed evaluation; this will mean we don’t end up reparsing everything again when things change.

There are only a handful of properties that impact parsing. You can see them on the ParseOptions type at https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.parseoptions, and the derived types which also have the language version. So the ones that would probably matter for getting this right (or close enough) would be to pass in:

•	/define (here’s how it’s processed: https://github.com/dotnet/roslyn/blob/7714ab4a89129ccc1885f961c0d3a12adce2add5/src/Compilers/Core/MSBuildTask/Csc.cs#L398) 
•	/doc (just passed in directly from DocumentationFile)
•	/features (here’s how it’s processed: https://github.com/dotnet/roslyn/blob/7714ab4a89129ccc1885f961c0d3a12adce2add5/src/Compilers/Core/MSBuildTask/ManagedCompiler.cs#L834-L848)
•	/langver (it’s just passed in directly from LangVersion)

The handling for /define is mostly handling some strange or error cases – simply just treating it as a string.Split/Join is probably a good enough of an approximation for 99% of projects. The handling for /doc is a bit interesting: we translate the path existing or not into the parsing mode on the ParseOptions type, which is why you don’t see the string there but it’s something we’d need to deal with.

I think it should also be possible to avoid reparsing trees that don’t have any #defines if only the /define switch changes; that would help speed up Debug/Release transitions if the project system wasn’t destroying and re-creating contexts.

**Comment by jasonmalinowski** [comment]
@drewnoakes I think it should be more than expressive enough as most of the transformations are just string splitting. As far as the project system side, is following that other PR good enough if I wanted to try getting the end-to-end working, or would it need something more complicated?

**Comment by drewnoakes** [comment]
@jasonmalinowski yes I believe that should suffice. You would need to plumb the property through the rule, dataflow, and into the data object used in that method.

**Comment by jaredpar** [comment]
@chsienki, @RikkiGibson

**Comment by chsienki** [comment]
Seems like a good idea. How does the project system actually get the properties to pass to the IDE? Does it evaluate the values of the CSC task and pass those in, or does the actual CSC build task get executed?

**Comment by jasonmalinowski** [comment]
@chsienki @RikkiGibson The actual Csc build task is executed, but with the switch that prevents it from actually invoking csc.exe or the compiler server. Essentially the loading is broken down into two steps:

1. The MSBuild evaluation pass, where we're only evaluating properties and item groups.
2. The MSBuild execution pass where we're actually running tasks/targets.

The project system starts telling Roslyn about information after the first pass, but since we don't have any parse options we're parsing with default options. Once the second pass happens we now get options, and potentially start parsing stuff over again.

**Comment by chsienki** [comment]
@jasonmalinowski Ok, my suggestion was going to be that we could do the command line 'construction' in the MSBuild task itself and return an out property, rather than having to do it via MSBuild evaluation, but it seems like this needs to happen *before* that point, so wouldn't work.

This is only really needed for the C# stuff in the IDE right? Xaml builds and friends won't need to use it? We already have various copies of the targets in different places that are slightly out of sync and I want to make sure we aren't going to make that worse.

**Comment by jasonmalinowski** [comment]
@chsienki Indeed, that's how it already works, but not until the execution pass. Because the execution pass is slow(er) than just the evaluation pass, that's ran async in the IDE.

> This is only really needed for the C# stuff in the IDE right? Xaml builds and friends won't need to use it? We already have various copies of the targets in different places that are slightly out of sync and I want to make sure we aren't going to make that worse.

Hmm, where are these other copies? That's news to me.

**Comment by chsienki** [comment]
Theres one in MSBuild that's part of `XamlPreCompile` https://github.com/dotnet/msbuild/blob/eac68aa8b922bb1d4c661c14138019f064565f7d/src/Tasks/Microsoft.CSharp.CurrentVersion.targets#L168

There are also razor components / compilations https://github.com/dotnet/sdk/blob/cc2a9c45df8bbfae6cc55a146b14dec93139b224/src/RazorSdk/Targets/Microsoft.NET.Sdk.Razor.Component.targets https://github.com/dotnet/sdk/blob/a30e465a2e2ea4e2550f319a2dc088daaafe5649/src/RazorSdk/Targets/Microsoft.NET.Sdk.Razor.Compilation.targets

And I think at least WinUI is doing the same thing.

**Comment by jasonmalinowski** [comment]
@chsienki So if those are copies of the task/target that's fine. This would only be consumed by the project system, and produced in the main targets file. Those other ones don't need any adjustment.

**Comment by jasonmalinowski** [comment]
@drewnoakes OK if I move this bug to Roslyn since at this point I think we're planning on doing most the work at this point?

**Comment by drewnoakes** [comment]
@jasonmalinowski no concerns with that, thanks. Feel free to file a new issue here with anything needed from our side and assign it to me.
